clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreAwarePartitionWiseKafkaConsumerService.java

```diff
@@ -0,0 +1,200 @@
+      consumerToLoadMap.put(i, overallLoad);
+    }
+
+    // Sort consumer by computed load.
+    Optional<Map.Entry<Integer, Long>> leastLoadedEntry =
+        consumerToLoadMap.entrySet().stream().min(Map.Entry.comparingByValue());
+    if (!leastLoadedEntry.isPresent()) {
+      throw new IllegalStateException(
+          "Unable to find least loaded consumer entry. Current load map: " + consumerToLoadMap);
+    }
+
+    int index = leastLoadedEntry.get().getKey();
+    long load = leastLoadedEntry.get().getValue();
+
+    if (load == Long.MAX_VALUE) {
+      throw new IllegalStateException("Did not find a suitable consumer with valid load.");
+    }
+
+    SharedKafkaConsumer consumer = getConsumerToConsumptionTask().getByIndex(index).getKey();
+    // Update resource identifier to consumer load map.
+    Map<PubSubConsumerAdapter, Integer> consumerMap = getResourceIdentifierToConsumerMap()
+        .computeIfAbsent(resourceIdentifier, key -> new VeniceConcurrentHashMap<>());
+    int existCount = consumerMap.getOrDefault(consumer, 0);
+    consumerMap.put(consumer, existCount + 1);
+
+    // Update RT topic partition consumer map.
+    if (topicPartition.getPubSubTopic().isRealTime()) {
+      getRtTopicPartitionToConsumerMap().computeIfAbsent(topicPartition, key -> new HashSet<>()).add(consumer);
+    }
+
+    getLOGGER().info(
+        "Picked consumer id: {}, assignment size: {}, computed load: {} for topic partition: {}, version topic: {}",
+        index,
+        consumer.getAssignmentSize(),
+        load,
+        topicPartition,
+        versionTopic);
+    return consumer;
+  }
+
+  @Override
+  void handleUnsubscription(
+      SharedKafkaConsumer consumer,
+      PubSubTopic versionTopic,
+      PubSubTopicPartition pubSubTopicPartition) {
+    super.handleUnsubscription(consumer, versionTopic, pubSubTopicPartition);
+
+    // Update resource identifier to consumer assignment map.
+    String resourceIdentifier = getResourceIdentifier(versionTopic);
+    if (!getResourceIdentifierToConsumerMap().containsKey(resourceIdentifier)) {
+      throw new IllegalStateException(
+          "Resource identifier consumer map does not contain expected resource identifier: " + resourceIdentifier);
+    }
+    int count = getResourceIdentifierToConsumerMap().get(resourceIdentifier).getOrDefault(consumer, 0);
+    if (count <= 0) {
+      throw new IllegalStateException("Consumer assignment count map does not contain matching consumer: " + consumer);
+    }
+    if (count == 1) {
+      getResourceIdentifierToConsumerMap().get(resourceIdentifier).remove(consumer);
+      if (getResourceIdentifierToConsumerMap().get(resourceIdentifier).isEmpty()) {
+        getResourceIdentifierToConsumerMap().remove(resourceIdentifier);
+      }
+    } else {
+      getResourceIdentifierToConsumerMap().get(resourceIdentifier).put(consumer, count - 1);
+    }
+  }
```

Comment:
I see a few problems here:
1. This function is not synchronized and it can be invoked concurrently and also `pickConsumerForPartition` would read the variables being updated here, so how can we guarantee correctness?
2. There are duplicate lookups here, and we should lookup once and continue to use it throughout the whole function.

Reply:
Good point, I missed unsubscribe path. I only looked at subscribe path and think that's good enough. Updated.